go-controller/pkg/clustermanager/network_cluster_controller.go

```diff
@@ -204,6 +220,79 @@ func (ncc *networkClusterController) init() error {
 	return nil
 }
 
+// updateNetworkStatus allows to report an status for networkClusterController's network via a UDN status condition
+// of type "NetworkAllocationSucceeded", if the network was created by UDN.
+// When at least one node reports an error, condition will be set to false and an event with node-specific error will be
+// generated.
+// Call this function after every node event handling, set errorMsg to empty string to report no error.
+func (ncc *networkClusterController) updateNetworkStatus(nodeName string, errorMsg string) {
+	if ncc.statusReporter == nil {
+		return
+	}
+
+	ncc.nodeErrorsLock.Lock()
+	defer ncc.nodeErrorsLock.Unlock()
+	if ncc.nodeErrors[nodeName] == errorMsg {
+		// error message didn't change for that node, no need to update
+		return
+	}
+	errsBefore := len(ncc.nodeErrors)
+	if errorMsg == "" {
+		delete(ncc.nodeErrors, nodeName)
+	} else {
+		ncc.nodeErrors[nodeName] = errorMsg
+	}
+
+	netName := ncc.NetInfo.GetNetworkName()
+	condition := getNetworkAllocationUDNCondition(len(ncc.nodeErrors) == 0)
+	events := make([]*util.EventDetails, 0, 1)
+	if errorMsg != "" {
+		events = append(events, &util.EventDetails{
+			EventType: util.EventTypeWarning,
+			Reason:    "NetworkAllocationFailed",
+			Note:      fmt.Sprintf("Error occurred for node %s: %s", nodeName, errorMsg),
+		})
+	}
+	if errsBefore == 0 && len(ncc.nodeErrors) == 0 || errsBefore > 0 && len(ncc.nodeErrors) > 0 {
+		// condition didn't change, no need to update
+		condition = nil
+	}
```

Comment:
hmmm

Not sure how I feel about reporting events without updating the condition. I feel that if the events are related to the condition, it should be updated.

Are you sure about it?

Reply:
I think it depends on the condition: if you have some special condition reason for a given error, then it should change with the error, but if the reason is generic like "InternalError", then I don't think updating it makes sense, since the condition itself didn't actually change.

Reply:
yeah that is exactly what I thought and I removed the extra logic from the UDN interface skipping condition updates.
But getting back to the your example, I don't know if there is a way to solve this, as you kinda want to not report same error if nothing changed in the object spec, but do report an error if something did to say to the user that it didn't help.. also doesn't feel quite right, but if I was about to make it work by comparing obj spec, I would try to add this logic to the events generation not to the condition update.

Reply:
you example made me think about 2 things:
1. events may get lost of expire, but then you are only left with message of "some node error" without even knowing which node you should fix
2. when there are multiple failed nodes, and you have fixed one of them, how do you know it actually worked? One way could be to report success events when an error is fixed or you could list all failed node names in the condition itself, and you don't need a success event.

Getting into account that reporting failed nodes in a condition message will solve both problems, I think I will add it.

Reply:
> as you kinda want to not report same error if nothing changed in the object spec

This would be the normal behavior of a controller for a resource. But this is not a normal controller for a resource. We consider the resource reconciles in a distributed manner and that's why we have error reports from different controller for that same resource.

This means that changing something in a different place than the resource spec itself may fix the error that we are reporting on that resource (or not). So that means we may want to report even if nothing changed on the resource itself.

> I don't know if there is a way to solve this

This would be my take: we need to be smart on when we send the status report from the source but not try to be very smart from the middleware to cache and filter out on no state change.

However, while we report for each an every error from the source (source I mean caller of `updateNetworkStatus`) then better to be caching in the middleware (meaning the implementation of `updateNetworkStatus`